### ISSUE TITLE
GPU utility is low

### ISSUE THREAD

**Issue by chenchen718** [ppwwyyxx/tensorpack#447]
When training, my GPU utility is low --- lower than 50% on average while the queue size is nearly 50. You mentioned in one post that it might be because some codes are running on CPU rather than GPU. How can I make sure that everything is running on GPU? Do I have to put every function for building the graph inside the class 'Model(ModelDesc)'?

I am using two GPUs for parallel training. What is wired is they seem to be utilized alternately --- when one GPU is 100%, the other is 0%; when one is 30%, the other is around 70%. What might be the reason?

Thanks a lot!

**Comment by ppwwyyxx** [comment]
Make sure that `_build_graph` build the part of graph from input to cost.

For your case, it could be your model has very different communication-computation ratio so it spends more time in communication. Or you might mistakenly introduce some dependencies among the two GPUs so they don't run together. It's best if you could show some code or at least describe what you've done.

**Comment by chenchen718** [comment]
@ppwwyyxx 
I independently tested the codes you provide in '[efficient DataFlow](http://tensorpack.readthedocs.io/en/latest/tutorial/efficient-dataflow.html)' as following
`ds = dataset.ILSVRC12(datadir, 'train', meta_dir=metadir, shuffle=True)
ds = AugmentImageComponent(ds, augmentors)
ds = PrefetchDataZMQ(ds, nr_proc=25)
ds = BatchData(ds, 256)`

and 
`ds = LMDBData(datadir, shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = PrefetchData(ds, 5000, 1)
ds = LMDBDataPoint(ds)
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
ds = AugmentImageComponent(ds, augmentors)
ds = PrefetchDataZMQ(ds, 25)
ds = BatchData(ds, 256, use_list=True)
TestDataSpeed(ds, 1000).start()`

The former one gives the performance of 0.43it/s and the latter one 4.27it/s. Then I used the two versions of reading data in the ['ResNet example'](https://github.com/ppwwyyxx/tensorpack/tree/master/examples/ResNet), I got the performances of  3.62it/s and 3.97it/s in 1-gpu case, and 1.43it/s and 1.42it/s in 2-gpu case. 

I don't understand why this happens. It seems that the use of GPU is influencing the data reading, and that for random read and sequential read, GPU is influencing in different ways.

**Comment by chenchen718** [comment]
One more question: are `InferenceRunner/QueueInput/queue_size` and `QueueInput/queue_size` the same thing?  `InferenceRunner/QueueInput/queue_size` seems to be always 50, but `QueueInput/queue_size` sometimes is very low.

**Comment by chenchen718** [comment]
Information of my two GPUs:

```
2017-10-22 17:07:07.234903: I tensorflow/core/common_runtime/gpu/gpu_device.cc:955] Found device 0 with properties: 
name: GeForce GTX 1080 Ti
major: 6 minor: 1 memoryClockRate (GHz) 1.582
pciBusID 0000:03:00.0
Total memory: 10.91GiB
Free memory: 10.13GiB
2017-10-22 17:07:07.765882: W tensorflow/stream_executor/cuda/cuda_driver.cc:523] A non-primary context 0x57d1530 exists before initializing the StreamExecutor. We haven't verified StreamExecutor works with that.
2017-10-22 17:07:07.767095: I tensorflow/core/common_runtime/gpu/gpu_device.cc:955] Found device 1 with properties: 
name: Quadro K6000
major: 3 minor: 5 memoryClockRate (GHz) 0.9015
pciBusID 0000:04:00.0
Total memory: 11.17GiB
Free memory: 11.10GiB
```

**Comment by ppwwyyxx** [comment]
The two queues are for training and validation respectively. Queues being empty basically says your data is slow.
When data is the bottleneck you should've seen the performance goes roughly from 3.6 -> 1.8 when goes from 1 GPU to 2.

I don't understand what have you done. Which "two versions" have you used in the end exactly? How could you possibly get 3.6it/s when data is only 0.4it/s.

**Comment by chenchen718** [comment]
By two versions, I mean 'random read' and 'sequential read' as shown in the two code blocks. I change nothing except for the `get_imagenet_dataflow` function in `imagenet_resnet_utils.py`. I try using these two versions to read the training data, and get the results. 'Random read' becomes faster and 'Sequential read' becomes slower.

**Comment by ppwwyyxx** [comment]
Become faster from 0.4it/s to 3.6it/s, with the same batch size 256? This is impossible and you probably read the number too early when it's unstable.

**Comment by chenchen718** [comment]
I am sorry. I just found that I was using different batch sizes. After correcting the batch size, I have the performances of 4.35 it/s and 28.31 it/s for 'random read' and 'sequential read', respectively. Now it seems that when training using GPUs, the performance is dropping dramatically. I think my GPUs are fine to run fast enough. They shouldn't become such a big bottleneck.

**Comment by ppwwyyxx** [comment]
You mean drop from 4.3 to 3.6? 3.6 sounds like a reasonable number for GTX 1080. You can check your queue size and GPU utilization.

**Comment by chenchen718** [comment]
Dropping from 4.3 to 3.6 sounds reasonable. But for the 'sequential read' case, it drops from 28.31 to 3.97, which is too big.  The data should be enough in this case, but when I use 2 GPUs, the performance drop from 3.97 (1 gpu) to 1.42. I don't expect it to be slower with 2 GPUs.

**Comment by chenchen718** [comment]
Queue size is 50 and GPU utilization is around 96%.

**Comment by ppwwyyxx** [comment]
It drops from 28.3 to 3.9 because that's about best your GPU can do. Yes it drops a lot but what else could you expect?
With a 1080Ti I can only get 2.7it/s for resnet18 batch=256. Don't even know how you could get 3.x.

If data is fast enough and you didn't modify the code about batch size, you should see about linear speed up. In no way it should drop and I still doubt whether you have measured things correctly.
If you still can't figure out, please attach the modification you've done and training log for at least two epochs with a small epoch size (like steps_per_epoch=50).

**Comment by chenchen718** [comment]
Sorry I forgot to mention that the results above are for resnet 50 with batch size 32. 

What I Modify is to replace 
```python
        ds = dataset.ILSVRC12(datadir, name, shuffle=True)
        ds = AugmentImageComponent(ds, augmentors, copy=False)
        ds = PrefetchDataZMQ(ds, cpu)
        ds = BatchData(ds, batch_size, remainder=False)
```
in the function `get_imagenet_dataflow` with 
```python 
        ds = LMDBData(datadir, shuffle=False)
        ds = LocallyShuffleData(ds, 50000)
        ds = PrefetchData(ds, 5000, 1)
        ds = LMDBDataPoint(ds)
        ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
        ds = AugmentImageComponent(ds, augmentors)
        ds = PrefetchDataZMQ(ds, 25)
        ds = BatchData(ds, batch_size)\
```
And switch off validation in training process.

The following are for resnet 18 with batchsize 256. This time I use two same 1080 ti GPUs instead of two different ones, and they seem to be collaborating well as you said. 

1. Use 1 1080 ti GPU;
```python
[1023 10:11:52 @base.py:232] Start Epoch 1 ...
100%|##########|50/50[00:37<00:00, 1.32it/s]
[1023 10:12:30 @base.py:242] Epoch 1 (global_step 50) finished, time:38.01 sec.
[1023 10:12:30 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-50.
[1023 10:12:30 @monitor.py:359] GPUUtil/0: 83.667
[1023 10:12:30 @monitor.py:359] QueueInput/queue_size: 47.456
[1023 10:12:30 @monitor.py:359] l2_regularize_loss: 0.48821
[1023 10:12:30 @monitor.py:359] learning_rate: 0.1
[1023 10:12:30 @monitor.py:359] train-error-top1: 0.89975
[1023 10:12:30 @monitor.py:359] train-error-top5: 0.68685
[1023 10:12:30 @monitor.py:359] xentropy-loss: 3.554
[1023 10:12:30 @base.py:232] Start Epoch 2 ...
100%|##########|50/50[00:31<00:00, 1.58it/s]
[1023 10:13:02 @base.py:242] Epoch 2 (global_step 100) finished, time:31.59 sec.
[1023 10:13:02 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-100.
[1023 10:13:02 @monitor.py:359] GPUUtil/0: 95.267
[1023 10:13:02 @monitor.py:359] QueueInput/queue_size: 49.818
[1023 10:13:02 @monitor.py:359] l2_regularize_loss: 0.48518
[1023 10:13:02 @monitor.py:359] learning_rate: 0.1
[1023 10:13:02 @monitor.py:359] train-error-top1: 0.89762
[1023 10:13:02 @monitor.py:359] train-error-top5: 0.67561
[1023 10:13:02 @monitor.py:359] xentropy-loss: 3.7917
[1023 10:13:02 @base.py:232] Start Epoch 3 ...
100%|##########|50/50[00:31<00:00, 1.58it/s]
[1023 10:13:34 @base.py:242] Epoch 3 (global_step 150) finished, time:31.68 sec.
[1023 10:13:34 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-150.
[1023 10:13:34 @monitor.py:359] GPUUtil/0: 94.733
[1023 10:13:34 @monitor.py:359] QueueInput/queue_size: 49.986
[1023 10:13:34 @monitor.py:359] l2_regularize_loss: 0.48217
[1023 10:13:34 @monitor.py:359] learning_rate: 0.1
[1023 10:13:34 @monitor.py:359] train-error-top1: 0.89215
[1023 10:13:34 @monitor.py:359] train-error-top5: 0.68439
[1023 10:13:34 @monitor.py:359] xentropy-loss: 4.0553
[1023 10:13:34 @base.py:232] Start Epoch 4 ...
100%|##########|50/50[00:30<00:00, 1.63it/s]
[1023 10:14:05 @base.py:242] Epoch 4 (global_step 200) finished, time:30.64 sec.
  0%|          |0/50[00:00<?,?it/s][1023 10:14:05 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-200.
[1023 10:14:05 @monitor.py:359] GPUUtil/0: 95.069
[1023 10:14:05 @monitor.py:359] QueueInput/queue_size: 49.999
[1023 10:14:05 @monitor.py:359] l2_regularize_loss: 0.47961
[1023 10:14:05 @monitor.py:359] learning_rate: 0.1
[1023 10:14:05 @monitor.py:359] train-error-top1: 0.8763
[1023 10:14:05 @monitor.py:359] train-error-top5: 0.64162
[1023 10:14:05 @monitor.py:359] xentropy-loss: 4.0434
[1023 10:14:05 @base.py:232] Start Epoch 5 ...
[1023 10:14:34 @base.py:242] Epoch 5 (global_step 250) finished, time:29.51 sec.
100%|##########|50/50[00:29<00:00, 1.69it/s]
[1023 10:14:35 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-250.
[1023 10:14:35 @monitor.py:359] GPUUtil/0: 94.5
[1023 10:14:35 @monitor.py:359] QueueInput/queue_size: 50
[1023 10:14:35 @monitor.py:359] l2_regularize_loss: 0.47672
[1023 10:14:35 @monitor.py:359] learning_rate: 0.1
[1023 10:14:35 @monitor.py:359] train-error-top1: 0.86738
[1023 10:14:35 @monitor.py:359] train-error-top5: 0.64623
[1023 10:14:35 @monitor.py:359] xentropy-loss: 3.7505
```

2. Use two 1080ti GPUs:
```python
100%|##########|50/50[00:26<00:00, 1.86it/s]
[1023 10:07:56 @base.py:242] Epoch 1 (global_step 50) finished, time:26.88 sec.
[1023 10:07:56 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-50.
[1023 10:07:57 @monitor.py:359] GPUUtil/0: 62.16
[1023 10:07:57 @monitor.py:359] GPUUtil/1: 54.36
[1023 10:07:57 @monitor.py:359] QueueInput/queue_size: 49.353
[1023 10:07:57 @monitor.py:359] l2_regularize_loss: 0.48833
[1023 10:07:57 @monitor.py:359] learning_rate: 0.1
[1023 10:07:57 @monitor.py:359] train-error-top1: 0.91258
[1023 10:07:57 @monitor.py:359] train-error-top5: 0.70073
[1023 10:07:57 @monitor.py:359] xentropy-loss: 3.5975
[1023 10:07:57 @base.py:232] Start Epoch 2 ...
100%|##########|50/50[00:16<00:00, 3.09it/s]
[1023 10:08:13 @base.py:242] Epoch 2 (global_step 100) finished, time:16.19 sec.
[1023 10:08:13 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-100.
[1023 10:08:13 @monitor.py:359] GPUUtil/0: 88.867
[1023 10:08:13 @monitor.py:359] GPUUtil/1: 79.067
[1023 10:08:13 @monitor.py:359] QueueInput/queue_size: 49.5
[1023 10:08:13 @monitor.py:359] l2_regularize_loss: 0.48509
[1023 10:08:13 @monitor.py:359] learning_rate: 0.1
[1023 10:08:13 @monitor.py:359] train-error-top1: 0.9074
[1023 10:08:13 @monitor.py:359] train-error-top5: 0.70409
[1023 10:08:13 @monitor.py:359] xentropy-loss: 3.8236
[1023 10:08:13 @base.py:232] Start Epoch 3 ...
  0%|          |0/50[00:00<?,?it/s][1023 10:08:28 @base.py:242] Epoch 3 (global_step 150) finished, time:15.10 sec.
100%|##########|50/50[00:15<00:00, 3.31it/s]
[1023 10:08:28 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-150.
[1023 10:08:28 @monitor.py:359] GPUUtil/0: 93.214
[1023 10:08:28 @monitor.py:359] GPUUtil/1: 89
[1023 10:08:28 @monitor.py:359] QueueInput/queue_size: 49.648
[1023 10:08:28 @monitor.py:359] l2_regularize_loss: 0.4818
[1023 10:08:28 @monitor.py:359] learning_rate: 0.1
[1023 10:08:28 @monitor.py:359] train-error-top1: 0.89979
[1023 10:08:28 @monitor.py:359] train-error-top5: 0.70344
[1023 10:08:28 @monitor.py:359] xentropy-loss: 4.0616
[1023 10:08:28 @base.py:232] Start Epoch 4 ...
100%|##########|50/50[00:14<00:00, 3.34it/s]
[1023 10:08:43 @base.py:242] Epoch 4 (global_step 200) finished, time:14.98 sec.
[1023 10:08:43 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-200.
[1023 10:08:43 @monitor.py:359] GPUUtil/0: 93.071
[1023 10:08:43 @monitor.py:359] GPUUtil/1: 94.214
[1023 10:08:43 @monitor.py:359] QueueInput/queue_size: 49.676
[1023 10:08:43 @monitor.py:359] l2_regularize_loss: 0.47896
[1023 10:08:43 @monitor.py:359] learning_rate: 0.1
[1023 10:08:43 @monitor.py:359] train-error-top1: 0.89005
[1023 10:08:43 @monitor.py:359] train-error-top5: 0.67852
[1023 10:08:43 @monitor.py:359] xentropy-loss: 4.1054
[1023 10:08:43 @base.py:232] Start Epoch 5 ...
  0%|          |0/50[00:00<?,?it/s][1023 10:09:01 @base.py:242] Epoch 5 (global_step 250) finished, time:17.26 sec.
100%|##########|50/50[00:17<00:00, 2.90it/s]
[1023 10:09:01 @saver.py:90] Model saved to train_log/imagenet-resnet-d18/model-250.
[1023 10:09:01 @monitor.py:359] GPUUtil/0: 96.25
[1023 10:09:01 @monitor.py:359] GPUUtil/1: 76.938
[1023 10:09:01 @monitor.py:359] QueueInput/queue_size: 49.638
[1023 10:09:01 @monitor.py:359] l2_regularize_loss: 0.47608
[1023 10:09:01 @monitor.py:359] learning_rate: 0.1
[1023 10:09:01 @monitor.py:359] train-error-top1: 0.87799
[1023 10:09:01 @monitor.py:359] train-error-top5: 0.66664
[1023 10:09:01 @monitor.py:359] xentropy-loss: 3.796
```

The data reading and processing speed  is 4.27it/s. This time, the speed performances all seem reasonable, isn't?

But the GPU utility is still not full. Is there a good way to increase GPU utility?

**Comment by ppwwyyxx** [comment]
As I said when you have about 95% utilization, that's the best you can get.

I don't know about why your 1080Ti seems slower. I would doubt it's a version issue. I'm using the latest of TF/cuda/cudnn/nvidida driver, if that matters. Now there seems to be no problems with the use of tensorpack.

**Comment by chenchen718** [comment]
I'll figure out whether I can work the GPU up somehow. Your suggestion really helps!

I have one more question regarding building my own network. Can I put the functions for network layers outside `class Model(ModelDesc)` and use a function in `_build_graph` to call them?

**Comment by ppwwyyxx** [comment]
You can. That's what the resnet example is doing.

**Comment by chenchen718** [comment]
Thank you very much for your kind help!

**Comment by chenchen718** [comment]
@ppwwyyxx  By "latest of TF/cuda/cudnn/nvidida drive", do you mean TF14/cuda9/cudnn7/nvidia384?

**Comment by ppwwyyxx** [comment]
Actually nvidia 387. But I think this doesn't matter too much.
TF1.4 has to be compiled from source to work with cuda9/cudnn7.
